[ALLUXIO-2564] Improve entry/exit debug logging

calvinjia opened the pull request:
https://alluxio.atlassian.net/browse/ALLUXIO-2564

Follow up to #4798 which makes the logging more failure resilient.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/13504/
Test PASSed.

apc999 commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -39,10 +39,35 @@
       return callable.call();
     } catch (AlluxioException e) {
-      LOG.debug("Internal Alluxio error when running rpc", e);
+      LOG.debug("Internal Error: {}", callable, e);
```
can we implement `public static <T> T call(RpcCallable<T> callable)` 
by `public static <T> T call(Logger logger, RpcCallable<T> callable)`?

jsimsa replied:
I think this block violates our exception handling guidelines, which are:

either log the exception or re-throw it but not both

@aaudiber can you please comment

calvinjia replied:
We need to log the exception here or the exception will only be thrown to the client and we will lose the server side stack trace.

jsimsa replied:
I see, makes sense ... in that case, disregard all the instances of this comment

aaudiber replied:
This case is unique in that it's the last place where we see the exception before it's sent over the wire. Client logs can be harder to track down, so at debug level I think it's reasonable to log and throw here.

aaudiber replied:
"Internal Error" suggests that something has gone wrong in the server, but this may not be the case, e.g. someone tries to create a file but it already exists. I'd prefer something more neutral like `Exception thrown by {}`.

calvinjia replied:
Would you be happier with the message if it was `Alluxio Error`? My thinking was to hint to the user which component is complaining (Alluxio or Non-Alluxio)

aaudiber replied:
I think we're making two unsafe assumptions here:
- The user understands where we draw the distinction between Alluxio and Non-Alluxio errors.
- `IOException` always means Non-Alluxio error and non-`IOException` always means Alluxio error.

I like Jiri's proposal of using
```java
("Exit (Error): {}", callable, e)
```
to indicate that an error has occurred.

apc999 commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -39,10 +39,35 @@
+   * @return the return value from calling the callable
+   * @throws AlluxioTException if the callable throws an Alluxio or runtime exception
+   */
+  public static <T> T call(Logger logger, RpcCallable<T> callable) throws AlluxioTException {
```
if this method is meant to be different from the other two overloading methods (in terms of what to log), perhaps use a different name. Overloading methods is supposed to be used when the logic is exactly the same.

jsimsa replied:
I would change the above method to invoke this method with the static LOG

apc999 commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -61,19 +86,51 @@
+   * @throws AlluxioTException if the callable throws an Alluxio or runtime exception
+   * @throws ThriftIOException if the callable throws an IOException
+   */
+  public static <T> T call(Logger logger, RpcCallableThrowsIOException<T> callable)
```
NIT: I would probably put `logger` as the second parameter. This is minor, up to you

jsimsa replied:
the single argument `call` method should invoke this one with `LOG`

calvinjia replied:
I think I will take Bin's suggestion and make two separate methods.

The reason why I put `logger` as the first parameter is to make the method call sites look cleaner in terms of style. If you think having `logger` as the first argument is counterintuitive even if this method is not overloaded, I'm fine with changing it.

The only reason why `logger` needs to be passed in the first place is to have the log message print the actual class ie. `FileSystemMasterClientServiceHandler` as opposed to `RPCUtils`. I think its valuable but I just want to point out why I did so in case you guys think it is not worth the complexity in `RPCUtils`.

The reason why I cannot just pass `LOG` as the default argument is because we do not want enter/exit debug level logging for certain RPCs like `sessionHeartbeat` since those will be far too frequent.

calvinjia replied:
If you guys think it is valuable to have enter/exit logging for periodic RPCs like `sessionHeartbeat`, we could remove the non-logging version all together. However I just didn't think it would be practical (ie. 100 clients connected to a worker means ~200 log lines per second).

jsimsa replied:
I agree that such excessive logging makes little sense ... if we wanted to get around that, we could supply a "nil" logger for RPCs we do not wish to log

calvinjia replied:
If you guys are fine with a hardcoded check `if (LOG != logger)`, I think that will be the simplest solution. This should have minimal performance overhead, but there will be some added to the critical path.

jsimsa replied:
works for me

jsimsa replied:
sg

apc999 replied:
sg

apc999 commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -39,10 +39,35 @@
       throw e.toThrift();
     } catch (Exception e) {
-      LOG.error("Unexpected error running rpc", e);
+      LOG.error("Unexpected Error: {}", callable, e);
```
do we expect `IOException` thrown here? should we have a separate catch block for `IOException` if yes?

jsimsa replied:
same comment as above

calvinjia replied:
In this version of `call`, `IOException` is not expected to be thrown, so it is treated as an `Unexpected Error`

aaudiber replied:
@apc999 this method is for rpcs which don't throw `IOException`

apc999 replied:
sg

gpang commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -39,10 +39,35 @@
+    try {
+      logger.debug("Enter: {}", callable);
+      T ret = callable.call();
+      logger.debug("Exit: {}", callable);
```
Do we also want timing information about the call? Not having to do manual timestamp subtraction would help debugging.

jsimsa replied:
do you want to put the "Exit" log statement into a finally block?

calvinjia replied:
I intended `Exit` to mean `Successful exit`, but I thought that was too verbose. We should see one `Enter` and one of `Exit/Error`. I thought seeing `Enter` then `Error` then `Exit` would be confusing.

jsimsa replied:
I have two concerns:

1) enabling automated log parsing requires a simple and consistent logging -- right now, our logging couples "Enter" with "Exit" or "Internal Error" or "I/O Error" or "Unexpected Error", which seems unnecessarily complicated

2) readability -- I find not using "Exit" upon encountering an exception confusing as the methods indeed exits ... someone not familiar with the above convention might look for "Exit" in the logs

How about using something along the following lines:

- `("Enter: {}", callable)` upon method entry
- `("Exit (OK): {}", callable)` upon clean method exit  
- `("Exit (Error): {}", callable, e)` upon unexpected method exit

calvinjia replied:
I like the `Exit (Error Type)` format, since we are explicitly categorizing the errors, I think it is valuable to propagate this to the user.

jsimsa replied:
sgtm

aaudiber replied:
+1

calvinjia replied:
Sure I can add timing info. There will be some performance overhead added to the critical path even without debug enabled, but if performance becomes an issue we can optimize afterward (ie. with `isDebugEnabled()` blocks).

calvinjia replied:
I talked with @aaudiber some more and our categorization is not enforced/consistent so for now I will just do `Exit (OK)` and `Exit (Error)`. I think its nice to have a qualifier so you can search for errors easily.

jsimsa replied:
sg

calvinjia replied:
@gpang @aaudiber On second thought, could we do this in a separate PR? This change is already pretty large and I would like to isolate the timing info change when evaluating any performance impact.

aaudiber replied:
sgtm

jsimsa commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -39,10 +39,35 @@
+      logger.debug("Exit: {}", callable);
+      return ret;
+    } catch (AlluxioException e) {
+      logger.debug("Internal Error: {}", callable, e);
```
again, I think we should either log the error or throw exception but not both

jsimsa commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -39,10 +39,35 @@
+      logger.debug("Internal Error: {}", callable, e);
+      throw e.toThrift();
+    } catch (Exception e) {
+      logger.error("Unexpected Error: {}", callable, e);
```
ditto

jsimsa commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -61,19 +86,51 @@
       return callable.call();
     } catch (AlluxioException e) {
-      LOG.debug("Internal Alluxio error when running rpc", e);
+      LOG.debug("Internal Error: {}", callable, e);
```
ditto

jsimsa commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -61,19 +86,51 @@
+      LOG.debug("Internal Error: {}", callable, e);
+      throw e.toThrift();
+    } catch (IOException e) {
+      LOG.debug("I/O Error: {}", callable, e);
```
ditto

aaudiber replied:
I think it would be fine to use the same "Exception running {}" for all of these exception logging statements. The name of the exception is already being logged along with the stack trace.

calvinjia replied:
I prefer having the category associated with the exception, so the user knows if there is something wrong within Alluxio, with I/O, or if something very bad has happened (unexpected). This information is hard to gather from just looking at the exception.

calvinjia replied:
Resolved offline, this isn't our current contract so I will just use `Exit (Error)`.

jsimsa commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -61,19 +86,51 @@
+      LOG.debug("I/O Error: {}", callable, e);
+      throw new ThriftIOException(e.getMessage());
+    } catch (Exception e) {
+      LOG.error("Unexpected Error: {}", callable, e);
```
ditto

jsimsa commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -61,19 +86,51 @@
+    try {
+      logger.debug("Enter: {}", callable);
+      T ret = callable.call();
+      logger.debug("Exit: {}", callable);
```
put into `finally` block?

jsimsa commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -61,19 +86,51 @@
       throw e.toThrift();
     } catch (IOException e) {
-      LOG.warn("I/O error when running rpc", e);
+      logger.debug("I/O Error: {}", callable, e);
```
log or throw exception but not both

jsimsa commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -61,19 +86,51 @@
       throw new ThriftIOException(e.getMessage());
     } catch (Exception e) {
-      LOG.error("Unexpected error running rpc", e);
+      logger.error("Unexpected Error: {}", callable, e);
```
ditto

jsimsa commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -61,19 +86,51 @@
   /**
-   * An interface representing a callable which can only throw Alluxio exceptions.
+   * An interface representing a callable which can only throw Alluxio exceptions. The toString
+   * method of classes implementing this interface should be in the format "CallName.
```
the `.` in the format string feels odd ... when I see `.`, I intuitively think "end of statement"

how about replacing `.` with `:` or alternatively `CallName(arg1:value1, arg2:value2, ...)`?

aaudiber replied:
+1

calvinjia replied:
Makes sense, I prefer to use `:`

jsimsa replied:
in that case, could we do:

`CallName: arg1=value1, arg2=value2, ...`

calvinjia replied:
sg

calvinjia commented:
@apc999 @aaudiber @gpang @jsimsa I've addressed most of the comments. For the timing info, I would prefer to do that in a separate PR since that will likely introduce performance overhead.

jsimsa commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -17,63 +17,103 @@
+
+  /**
+   * An interface representing a callable which can only throw Alluxio exceptions. The toString
+   * method of classes implementing this interface should be in the format "CallName.
```
update format string

jsimsa commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -88,7 +128,9 @@
   /**
-   * An interface representing a callable which can only throw Alluxio or IO exceptions.
+   * An interface representing a callable which can only throw Alluxio or IO exceptions. The
+   * toString method of classes implementing this interface should be in the format "CallName.
```
ditto

jsimsa commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -17,63 +17,103 @@
+  }
+
+  /**
+   * Calls the given {@link RpcCallableThrowsIOException} and handles any exceptions thrown. The
```
the javadoc for the format string seems to be duplicated here and in the RpcCallable interface ... I feel like we only need it in one of those places

calvinjia replied:
Good catch, I think it is only required in the `callAndLog` method now, since `RpcCallables` can use `call` without overriding `toString`

jsimsa commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -17,63 +17,103 @@
+   * @return the return value from calling the callable
+   * @throws AlluxioTException if the callable throws an Alluxio or runtime exception
+   */
+  public static <T> T callAndLog(Logger logger, RpcCallable<T> callable) throws AlluxioTException {
```
so it seems like the difference between `call` and `callAndLog` is that `call` only logs errors, while callAndLog logs the entry and exit ... but it is confusing that the error exit points are logged in the `call` method

would the following make more sense:

```
  public static <T> T call(Logger logger, boolean logControlFlow, RpcCallable<T> callable) throws AlluxioTException {
    try {
      if (logControlFlow) {
        logger.debug("Enter: {}", callable);
      }
      T ret = callable.call();
      if (logControlFlow) {
        logger.debug("Exit (OK): {}", callable);
      }
    } catch (AlluxioException e) {
      if (logControlFlow) {
        logger.debug("Exit (Error): {}", callable);
      }
      logger.debug("Internal Alluxio error ...", e);
      throw e.toThrift();
    } catch (Exception e) {
      if (logControlFlow) {
        logger.debug("Exit (Error): {}", callable);
      }
      logger.error("Unexpected error ...", e);
      throw new UnexpectedAlluxioException(e).toThrift();
    }
  }
```

the point of doing the above being to separate the control flow logging (which is conditional) from error logging

jsimsa replied:
my main concern of the current implementation is that it allows for "Exit" log statements that do not have matching "Enter" statements

calvinjia replied:
@aaudiber and I also talked about this. We thought `Exit (Error)` made sense even without a corresponding `Enter` statement. If debug logging is enabled, we will always get matching `Enter/Exit`, and if it is disabled, the user will only see `Exit` statements which is reasonable since we are exiting, just with an error.

The upside of your proposed approach is that there is only 1 method and we get `Enter/Exit` matching logs. The downside of that approach is the method is fairly complicated and will get more complicated as we want to enrich the additional logging.

jsimsa replied:
ok, I still find the API (`callAndLog` vs `call` which also logs) unintuitive, but my main concerns for this change have been addressed so if you and @aaudiber are happy with it, then I won't stand in the way of progress :-)

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/13520/
Test PASSed.

jsimsa commented on core/server/common/src/main/java/alluxio/RpcUtils.java:
```diff
@@ -17,63 +17,103 @@
   /**
-   * An interface representing a callable which can only throw Alluxio exceptions.
+   * Calls the given {@link RpcCallable} and handles any exceptions thrown. The callable should
+   * implement a toString with the following format: "CallName: arg1=value1, arg2=value2,...".
```
the javadoc for the format string seems to be duplicated here and in the RpcCallable interface ... I feel like we only need it in one of those places

jsimsa commented on core/server/master/src/main/java/alluxio/master/block/BlockMasterClientServiceHandler.java:
```diff
@@ -52,7 +56,7 @@ public long getServiceVersion() {
   @Override
   public List<WorkerInfo> getWorkerInfoList() throws AlluxioTException {
-    return RpcUtils.call(new RpcCallable<List<WorkerInfo>>() {
+    return RpcUtils.call(LOG, new RpcCallable<List<WorkerInfo>>() {
```
how do you decide when to use `call` vs `callAndLog`?

calvinjia replied:
Everything which is not automatic/periodic should eventually use `callAndLog`, currently I am only targeting the 3 most used service handlers and based on the usefulness of this approach, we can  expand the use to other handlers.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/13521/
Test PASSed.

ifcharming commented on core/server/master/src/main/java/alluxio/master/file/FileSystemMasterClientServiceHandler.java:
```diff
@@ -279,100 +305,122 @@ public String call() throws AlluxioException {
   @Override
   public void remove(final String path, final boolean recursive)
       throws AlluxioTException, ThriftIOException {
     LOG.debug("Enter Remove. path:{}, recursive:{}", path, recursive);
```
sorry for the post-merge comment: but shall we remove this `LOG.debug`?

calvinjia replied:
@ifcharming Good catch, I did another pass to clean this up in #4808

ParanoiaW commented on keyvalue/server/src/main/java/alluxio/master/keyvalue/KeyValueMasterClientServiceHandler.java:
```diff
@@ -51,7 +57,7 @@ public long getServiceVersion() {
   @Override
   public void completePartition(final String path, final PartitionInfo info)
       throws AlluxioTException {
```
Shall we override toString() for these methods?

calvinjia commented:
@ParanoiaW Yes, if using `callAndLog`, the callable should override `toString`
